python/lsst/pipe/base/_task_metadata.py

```diff
@@ -0,0 +1,271 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["TaskMetadata"]
+
+import warnings
+from collections.abc import Sequence
+from deprecated.sphinx import deprecated
+
+from typing import Dict, List, Union
+from pydantic import BaseModel, StrictInt, StrictFloat, StrictBool, StrictStr
+
+_DEPRECATION_REASON = "Will be removed after v25."
+_DEPRECATION_VERSION = "v24"
+
+
+def _isListLike(v):
+    return isinstance(v, Sequence) and not isinstance(v, str)
+
+
+class TaskMetadata(BaseModel):
+    """Dict-like object for storing task metadata.
+    Metadata can be stored at two levels: single task or task plus subtasks.
+    The later is called full metadata of a task and has a form
+
+        topLevelTaskName:subtaskName:subsubtaskName.itemName
+
+    Metadata item key of a task (`itemName` above) must not contain `.`,
+    which serves as a separator in full metadata keys and turns
+    the value into sub-dictionary. `KeyError` is raised if more than one
+    separator is detected in a key.
+
+    Deprecated methods are for compatibility with
+    the predecessor containers.
+    """
+
+    # Metadata is limited -- float must come before int
+    __root__: Dict[str, Union["TaskMetadata", StrictFloat, StrictInt, StrictBool, StrictStr,
+                              List[StrictFloat], List[StrictInt], List[StrictBool], List[StrictStr]]] = {}
+
+    def add(self, name, value):
+        """Store a new value, adding to a list if one already exists.
+
+        Parameters
+        ----------
+        name : `str`
+            Name of the metadata property.
+        value
+            Metadata property value.
+        """
+        if self.__contains__(name):
+            v = self.__getitem__(name)
+            if isinstance(v, TaskMetadata):
```

Comment:
Is there any way to get rid of the very large number of `isinstance()` calls in this class?  Perhaps maintaining separate top-level and second-level dictionaries, so an `in` check suffices?

Also maybe separate value and list storage so that can also avoid `isinstance()`?

Reply:
One motivation for this class was to try to make it as close to a Mapping as possible. Is performance an issue given what we are actually doing with this class (which is mostly calling `add()` and only having tens of keys.

If we don't care about the serialized JSON explicitly showing the scalar/list/TaskMetadata keys as distinct parts of the model then it probably will make it more efficient to do and I'll remove the `__root__` from the model.